ddtrace/sampler.py

```diff
@@ -50,50 +48,47 @@ def sample(self, span):
         return sampled
 
 
-def _key(service=None, env=None):
-    service = service or ''
-    env = env or ''
-    return 'service:' + service + ',env:' + env
-
-
-_default_key = _key()
-
-
 class RateByServiceSampler(object):
     """Sampler based on a rate, by service
 
     Keep (100 * `sample_rate`)% of the traces.
     The sample rate is kept independently for each service/env tuple.
     """
 
+    @staticmethod
+    def _key(service=None, env=None):
+        """Compute a key with the same format used by the Datadog agent API."""
+        service = service or ''
+        env = env or ''
+        return 'service:' + service + ',env:' + env
+
     def __init__(self, sample_rate=1):
-        self._lock = Lock()
-        self._by_service_samplers = {}
-        self._by_service_samplers[_default_key] = RateSampler(sample_rate)
+        self.sample_rate = sample_rate
+        self._by_service_samplers = self._get_new_by_service_sampler()
 
-    def _set_sample_rate_by_key(self, sample_rate, key):
-        with self._lock:
-            if key in self._by_service_samplers:
-                self._by_service_samplers[key].set_sample_rate(sample_rate)
-            else:
-                self._by_service_samplers[key] = RateSampler(sample_rate)
+    def _get_new_by_service_sampler(self):
+        return {
+            self._default_key: RateSampler(self.sample_rate)
+        }
 
     def set_sample_rate(self, sample_rate, service='', env=''):
-        self._set_sample_rate_by_key(sample_rate, _key(service, env))
+        self._by_service_samplers[self._key(service, env)] = RateSampler(sample_rate)
 
     def sample(self, span):
         tags = span.tracer().tags
         env = tags['env'] if 'env' in tags else None
-        key = _key(span.service, env)
-        with self._lock:
-            if key in self._by_service_samplers:
-                return self._by_service_samplers[key].sample(span)
-            return self._by_service_samplers[_default_key].sample(span)
+        key = self._key(span.service, env)
+        return self._by_service_samplers.get(
+            key, self._by_service_samplers[self._default_key]
+        ).sample(span)
 
     def set_sample_rate_by_service(self, rate_by_service):
+        new_by_service_samplers = self._get_new_by_service_sampler()
         for key, sample_rate in iteritems(rate_by_service):
-            self._set_sample_rate_by_key(sample_rate, key)
-        with self._lock:
-            for key in list(self._by_service_samplers):
-                if key not in rate_by_service and key != _default_key:
-                    del self._by_service_samplers[key]
+            new_by_service_samplers[key] = RateSampler(sample_rate)
+
+        self._by_service_samplers = new_by_service_samplers
```

Comment:
hmm, this is an interesting change, I am not sure we can do this, we might need to keep any existing service/env keys around.

I am not 100% sure the agent returns all previously known service/env keys... we'll need to check the implementation in the agent or ask @gbbr

Reply:
If I read the previous function correctly it:
- adds set the rates for every service returned by the agent
- deletes all the rates for every service not returned by the agent

So in the end, you have only the service returned by the agent in the `_by_service_samplers`. What do I miss?